src/main/java/mezz/jei/config/ClientConfig.java

```diff
@@ -36,6 +36,11 @@
 	private final ConfigValues values;
 	private List<? extends String> searchColors = Arrays.asList(ColorGetter.getColorDefaults());
 	public static final List<IngredientSortStage> ingredientSorterStagesDefault = Arrays.asList(
+		IngredientSortStage.WEAPON_DAMAGE,
```

Comment:
I think since there is probably some overhead to calculating these, they should not be enabled by default. I would only feel comfortable including these by default if we can get some information about the performance with vs without these sorting stages.

Reply:
I'll try to find a large 1.16 pack and give it a few goes.  What kind of information do you want, just some before vs after and a % delta?
Also, I didn't know I could change the sort order until I went in to figure out how to add this.  Need something to advertise it on the config panel at the very least.

Reply:
Testing done with ATM 6, 1.6.5

Using this to profile:
@Override
	public List<IIngredientListElement<?>> getIngredientList(String filterText) {
		filterText = filterText.toLowerCase();
		if (!filterText.equals(filterCached)) {
			List<IIngredientListElementInfo<?>> ingredientList = getIngredientListUncached(filterText);
			LoggedTimer filterTimer = new LoggedTimer();
			filterTimer.start("Filtering and Sorting: " + filterText);
			ingredientListCached = ingredientList.stream()
				.sorted(sorter.getComparator(this, this.ingredientManager))
				.map(IIngredientListElementInfo::getElement)
				.collect(Collectors.toList());
			filterTimer.stop();
			LogManager.getLogger().info("Filter has " + ingredientListCached.size() + " of " + ingredientList.size());
			filterCached = filterText;
		}
		return ingredientListCached;
	}

new JEI:

[01:46:17] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: ...
[01:46:18] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting:  took 1.252 s
[01:46:18] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 25784 of 25784
[01:46:29] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: a...
[01:46:30] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: a took 745.6 ms
[01:46:30] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 21412 of 21412
[01:46:41] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: al...
[01:46:41] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: al took 127.3 ms
[01:46:41] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 4266 of 4266
[01:46:41] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: all...
[01:46:41] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: all took 22.05 ms
[01:46:41] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 1054 of 1054
[01:46:50] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: al...
[01:46:50] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: al took 126.3 ms
[01:46:50] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 4266 of 4266
[01:46:52] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: a...
[01:46:53] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: a took 710.6 ms
[01:46:53] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 21412 of 21412
[01:46:54] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: ...
[01:46:55] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting:  took 878.6 ms
[01:46:55] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 25784 of 25784
-----------
New JEI, old default.
[01:54:56] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: ...
[01:54:56] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting:  took 35.01 ms
[01:54:56] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 25784 of 25784
[01:55:04] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: a...
[01:55:04] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: a took 21.29 ms
[01:55:04] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 21412 of 21412
[01:55:06] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: al...
[01:55:06] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: al took 4.855 ms
[01:55:06] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 4262 of 4262
[01:55:06] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: all...
[01:55:06] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: all took 1.282 ms
[01:55:06] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 1052 of 1052
[01:55:07] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: al...
[01:55:07] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: al took 5.232 ms
[01:55:07] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 4262 of 4262
[01:55:07] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: a...
[01:55:07] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: a took 20.01 ms
[01:55:07] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 21412 of 21412
[01:55:08] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: ...
[01:55:09] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting:  took 28.66 ms
[01:55:09] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 25784 of 25784

The difference is too great.  I'm going to try to add some form of caching to help.

Reply:
Thanks for checking. I am not sure if caching will help much, because it will still take a long time on the first sort when everything is uncached.

It would probably be better to just improve the visibility of the config so that people can add these if they want to.
We can also see if the code can be refactored so that the list is only sorted one time, and then we make sure the order is always preserved after filtering. I think I only filter every time you search right now because it was fast to do, so there was no reason to complicate it by optimizing that part.

Reply:
Two more stats:
-----------
New Sort with caching?
[17:14:56] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: ...
[17:14:58] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting:  took 1.636 s
[17:14:58] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 25784 of 25784
[17:15:05] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: a...
[17:15:06] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: a took 977.7 ms
[17:15:06] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 21411 of 21411
[17:15:06] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: al...
[17:15:06] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: al took 209.3 ms
[17:15:06] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 4260 of 4260
[17:15:06] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: all...
[17:15:06] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: all took 28.37 ms
[17:15:06] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 1051 of 1051
[17:15:10] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: al...
[17:15:10] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: al took 146.9 ms
[17:15:10] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 4260 of 4260
[17:15:10] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: a...
[17:15:11] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: a took 878.7 ms
[17:15:11] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 21411 of 21411
[17:15:12] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: ...
[17:15:13] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting:  took 1.163 s
[17:15:13] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 25784 of 25784

---
Newish sort with values in IngredientListElementInfo (First version I committed.):
[17:45:26] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: ...
[17:45:26] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting:  took 269.6 ms
[17:45:26] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 25784 of 25784
[17:45:38] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: a...
[17:45:38] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: a took 121.9 ms
[17:45:38] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 21412 of 21412
[17:45:38] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: al...
[17:45:38] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: al took 48.25 ms
[17:45:38] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 4263 of 4263
[17:45:38] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: all...
[17:45:38] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: all took 4.378 ms
[17:45:38] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 1050 of 1050
[17:45:39] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: al...
[17:45:39] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: al took 22.06 ms
[17:45:39] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 4263 of 4263
[17:45:39] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: a...
[17:45:39] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: a took 152.8 ms
[17:45:39] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 21412 of 21412
[17:45:39] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: ...
[17:45:39] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting:  took 179.9 ms
[17:45:39] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 25784 of 25784

So, at the nearly absolute best is about 6x slower.

Reply:
I was caching the wrong answer.  The next version should be pretty neat and always fast, except once.

Reply:
So, I have it sort the entire list, save that into an index field in the ElementInfo, then use that to do all future sorts.
After the first load, all filtering is faster than it was with the old sort, at the cost of 1 new integer in IngredientListElementInfo.
Caching the final index on first use:
[19:43:37] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: ...
[19:43:37] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Pre-Sorting....
[19:43:39] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Pre-Sorting. took 1.210 s
[19:43:39] [Render thread/INFO] [me.je.in.IngredientFilter/]: Sort has 25784
[19:43:39] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting:  took 1.236 s
[19:43:39] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 25784 of 25784
[19:43:48] [Server thread/INFO] [ne.mi.co.Co.apotheosis/COREMODLOG]: Patching CombatRules#getDamageAfterMagicAbsorb
[19:44:02] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: a...
[19:44:02] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: a took 6.936 ms
[19:44:02] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 21413 of 21413
[19:44:03] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: al...
[19:44:03] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: al took 1.244 ms
[19:44:03] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 4262 of 4262
[19:44:03] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: all...
[19:44:03] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: all took 363.1 μs
[19:44:03] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 1052 of 1052
[19:44:04] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: al...
[19:44:04] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: al took 1.151 ms
[19:44:04] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 4262 of 4262
[19:44:04] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: a...
[19:44:04] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: a took 7.384 ms
[19:44:04] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 21413 of 21413
[19:44:04] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting: ...
[19:44:04] [Render thread/INFO] [me.je.ut.LoggedTimer/]: Filtering and Sorting:  took 9.905 ms
[19:44:04] [Render thread/INFO] [me.je.in.IngredientFilter/]: Filter has 25784 of 25784

Reply:
Is there a way to change JEI's code so that this list is only ever sorted once, when it's created?
Adding 1.2s to JEI's startup time in a giant pack is fine if as long as filtering the ingredient list is always quick.

Reply:
No, I think if you use the pipe symbol to do multiple searches in one query, the order gets mixed up.  There is a need to do some sorting at the tail end.  I'm not certain, but that's my working theory based on what I saw.
As it is, just once per world load (or /reload) when you go into an inventory screen seems to be a pretty good place.  Also, if you edit one of the sorting files, /reload would let you see the new order.

Reply:
@mezz  Is there something more you are waiting for from me?

Reply:
I think this looks pretty good. We should force it to calculate the sort indexes some time at startup instead of on the first search, to avoid a lag when you type something. Other than that I think this is good to merge.

Reply:
I think it can't be done any earlier than near the end of world loading.  Some of the errors I've seen seem to to indicate that not everything is done figuring out its existence until then.  I may try to find the right spot then.  It just still needs to respect /reload for changes to the sort to be easy to test.

Reply:
JEI itself doesn't load until tags and recipes are synced from the server, during world load.
What here needs to be loaded later?

Reply:
I'm mostly learning how it all is working.
I basically need to make sure a sort happens after all of "[18:10:06] [Render thread/INFO] [me.je.in.IngredientManager/]: Ingredients are being added at runtime: 10 mezz.jei.plugins.jei.ingredients.DebugIngredient" are done.  
I'm just not quite sure when that is true.  I think I'm close enough though.